src/services/pool.ts

```diff
@@ -0,0 +1,112 @@
+import { Db, Filter, FindOptions, MongoClient, WithId } from 'mongodb';
+import { createMessage, initiateLogger } from './logger';
+import { assertTrailingSlash } from '../utils';
+
+/** BEGIN typing for DB */
+
+type EnvKeyedObject = {
+  prd: string;
+  preprd: string;
+  dotcomstg: string;
+  dotcomprd: string;
+};
+
+interface BranchEntry {
+  name: string;
+  gitBranchName: string;
+  active: boolean;
+  urlAliases: string[];
+  [key: string]: any;
+}
+
+interface RepoDocument extends WithId<Document> {
+  repoName: string;
+  project: string;
+  branches: BranchEntry[];
+  url: EnvKeyedObject;
+  prefix: EnvKeyedObject;
+  search: Record<string, string>;
+}
+
+interface BranchResponse {
+  gitBranchName: string;
+  active: boolean;
+  fullUrl: string;
+}
+
+interface RepoResponse {
+  repoName: string;
+  project: string;
+  search: Record<string, string>;
+  branches: BranchResponse[];
+}
+
+/** END typing for DB */
+
+const ATLAS_URI = process.env.ATLAS_URI ?? '';
+const REPOS_COLLECTION = 'repos_branches';
+const DB_NAME = process.env.POOL_DB_NAME ?? 'pool';
+const ENV_URL_KEY = (process.env.SNOOTY_ENV ?? 'dotcomprd') as keyof EnvKeyedObject;
+
+const logger = initiateLogger();
+const client = new MongoClient(ATLAS_URI);
+let dbInstance: Db | null = null;
+
+// exposes db instance
+export const poolDb = async () => {
+  if (dbInstance) return dbInstance;
+
+  try {
+    await client.connect();
+    dbInstance = client.db(DB_NAME);
+    return dbInstance;
+  } catch (e) {
+    logger.error(createMessage(`Error while connecting client: ${e}`));
+    throw e;
+  }
+};
+
+export const closeDBConnection = async () => {
+  await client.close();
+};
+
+export const findAllRepos = async (query: Filter<RepoDocument> = {}, options: FindOptions = {}, reqId?: string) => {
+  try {
+    const db = await poolDb();
+    const defaultSort: FindOptions = {
+      sort: { repoName: 1 },
+    };
+    const strictOptions: FindOptions = {
+      projection: {
+        repoName: 1,
+        project: 1,
+        branches: 1,
+        url: 1,
+        prefix: 1,
+      },
+    };
+    const findOptions = { ...defaultSort, ...options, ...strictOptions };
+    return await db.collection<RepoDocument>(REPOS_COLLECTION).find(query, findOptions).map(mapRepos).toArray();
```

Comment:
Wondering if we should have the find query omit certain repo names that we have specifically for content staging purposes only, such as the internal docs repo. Unsure if this is something we should create a static list for in the code to perform a `$nin` query, or if we should mark the documents in the db with a new field to query against like `contentStagingOnly` or `deployable`.

This could also be seen as out of scope, so happy to drop if not important rn. What do you think?

Reply:
i'd like to flesh this out as well. i think a static list works for initial run, and i think a clean up of all static lists in different repos could be addressed in a single ticket for the former (if it comes down to it)

also, sanity checking on repos without any active branches: i think these are "sunset/deprecated" ie. cloud-osb-docs but still should be returned, to provide info on deprecated projects/versions

Reply:
pushed a query filter to not regex match /internal/ (this omits `"repoName": "docs-mongodb-internal"` and `"repoName": "docs-mongodb-internal-base",`)

Reply:
> also, sanity checking on repos without any active branches: i think these are "sunset/deprecated" ie. cloud-osb-docs but still should be returned, to provide info on deprecated projects/versions

Unsure whether inactive branches and repos are needed by the chatbot team. I'm not sure what the value is of returning information on a product that the company no longer supports, but yeah, definitely good to check in on.

> pushed a query filter to not regex match /internal/ (this omits "repoName": "docs-mongodb-internal" and "repoName": "docs-mongodb-internal-base",)

Gotcha, I think something like this would work as part of the initial implementation